llvm/lib/Target/AIE/AIEMachineScheduler.cpp

```diff
@@ -883,8 +890,34 @@ bool AIEPreRASchedStrategy::isAvailableNode(SUnit &SU, SchedBoundary &Zone,
 
   // The node will likely cause a spill, only consider it schedule-able if
   // there is no pending node that can reduce the register pressure.
-  return findPressureReducer(WorstPC.getPSet(), Zone.Pending.elements(),
-                             BotRPT) == nullptr;
+  if (const SUnit *PendingPressureReducer = findPressureReducer(
+          WorstPC.getPSet(), Zone.Pending.elements(), BotRPT);
+      PendingPressureReducer && canBeDelayed(SU, *PendingPressureReducer)) {
+    LLVM_DEBUG(dbgs() << "** Delaying SU(" << SU.NodeNum << "): Waiting for SU("
+                      << PendingPressureReducer->NodeNum << ")\n");
+
+    // Keep track of PendingPressureReducer to avoid cycles of SUs
+    // delaying each other.
+    SUDelayerMap[SU.NodeNum] = PendingPressureReducer->NodeNum;
+    return false;
+  }
+
+  // Can't prove a pending SU will help reduce reg pressure, keep as available.
+  return true;
+}
+
+bool AIEPreRASchedStrategy::canBeDelayed(const SUnit &DelayedSU,
+                                         const SUnit &Delayer) const {
+  std::function<bool(unsigned)> Impl = [&](unsigned SUNum) {
+    if (SUNum == UnknownSUNum)
+      return false;
+    if (SUNum == DelayedSU.NodeNum)
+      return true;
+    return Impl(SUDelayerMap[SUNum]);
```

Comment:
use .at() here?

Reply:
This is actually an `std::vector` 🙈 I just use it as a flat map.

Reply:
std::vector has .at too, giving you bounds checks?